### DISCUSSION TITLE
Question about Value.IsConcrete()?

### DISCUSSION THREAD

**Discussion by verdverm** [cuelang/cue#884] [Q&A]
### What version of CUE are you using (`cue version`)?

`0.3.0`

### What did you do?

Called `cue.Value.IsConcrete()` via the Go API and got unexpected results. If this does indeed seem like a bug, then I can create a reproducer.

In the example below, calling `IsConcrete()` for `ex.bar.` reports true, is this expected?

```
#A: {
  a1: string
  a2: string | *""
}

#B: {
  foo: string
  bar: #A
}

ex: #B & {
  foo: "foo"
}
```

### What did you expect to see?

`ex.bar.IsConcrete()` returning **false**

### What did you see instead?

`ex.bar.IsConcrete()` returning **true**

**Comment by myitcv** [comment]
AFAIU, this is working "as expected" but is effectively the same problem of a list type defaulting to `[]`, picked up in the [required fields proposal](https://github.com/cuelang/cue/issues/822). i.e. the default value of `#A` here is `{}`. What I think you are looking to distinguish is whether `ex.bar` is specified or not, and that is where the `!` proposal comes in.

**Reply by verdverm** [reply]
I believe what I actually want here is `ex.bar.Validate(cue.Concrete(true))`. In short, I am searching for values which have not been specified yet. I'd like to make a list of them and print them out.

**Comment by mpvl** [comment]
It is working as expected because, as documented, a struct is always considered concrete. 

As per the doc:
> It does not verify that values of lists or structs are concrete themselves.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/884.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
